### PR TITLE
[JENKINS-64026] Only add git tag action if admin enables it

### PR DESCRIPTION
**Pull request by MarkEWaite**
## [JENKINS-64026](https://issues.jenkins-ci.org/browse/JENKINS-64026) - Only add git tag action if Administrator enables it

The git tag action was added many years ago based on the Subversion tag action.  It creates a tag in a workspace if the workspace is available.  It does not publish the tag to the upstream repository.  It does not assure that the workspace exists.  It does not assure that the tag is preserved in any location more permanent than a workspace.  It does not seem to function in workspaces that are on a Jenkins agent rather than the Jenkins controller.

Rather than waste space in the UI for an action with very little value, remove it by default and allow the administrator to enable it from the "Manage Jenkins" page.  This will no longer add the "No tags", "One tag", and "Multiple tags" actions to the Jenkins job result pages.  See the pictures below for examples of the user interface change.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] Documentation in README has been updated as necessary
- [x] Online help has been added and reviewed for any new or modified fields
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Further comments

Jenkins workspaces are frequently ephemeral.  An ephemeral workspace likely does not exist at the time the tag action attempts to create the tag.

The git tag action previously used a table based layout.  If div based layout is available, it will now use a div based layout.

Pull request also fixes a minor inconsistency in the credential hiding documentation while updating the documentation to explain the hide git tag action checkbox.

## Job Page with Tag Action

screencapture-mark-pc2-markwaite-net-8080-job-Bugs-Pipeline-Checks-job-jenkins-bugs-multibranch-pipeline-git-protocol-job-JENKINS-64000-26-2020-11-11-19_02_36-edit

## Job Page without Tag Action

screencapture-mark-pc2-markwaite-net-8080-job-Bugs-Pipeline-Checks-job-jenkins-bugs-multibranch-pipeline-git-protocol-job-JENKINS-64000-27-2020-11-11-19_03_41-edit

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -193,6 +193,19 @@ Preserve second fetch during initial checkout::
   This setting is only needed if there is a bug in the redundant fetch removal logic.
   If you enable this setting, please report a git plugin issue that describes why you needed to enable it.
 
+[[do-not-add-git-tag-action-to-jobs]]
+Do not add git tag action to jobs::
```
Comment (timja) [inline README.adoc:197]: could this be the default? I can't see people wanting to use this these days

File: README.adoc [hunk 2]
```diff
@@ -193,6 +193,19 @@ Preserve second fetch during initial checkout::
   This setting is only needed if there is a bug in the redundant fetch removal logic.
   If you enable this setting, please report a git plugin issue that describes why you needed to enable it.
 
+[[do-not-add-git-tag-action-to-jobs]]
```
Comment (timja) [inline README.adoc:196]: new screenshot looks way nicer 😄

File: src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly [hunk 3]
```diff
@@ -37,10 +37,106 @@ THE SOFTWARE.
...
+          <th>
+            <d:invokeBody/>
+          </th>
+        </j:otherwise>
+      </j:choose>
+    </d:tag>
+    <d:tag name="td">
+      <j:choose>
+        <j:when test="${divBasedFormLayout}">
+          <div class="td">
```
Comment (timja) [inline src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly:85]: tr is the only class that should actually on the migrated elements
```suggestion
          <div>
```

File: src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly [hunk 4]
```diff
@@ -37,10 +37,106 @@ THE SOFTWARE.
...
+          <tr>
+            <d:invokeBody/>
+          </tr>
+        </j:otherwise>
+      </j:choose>
+    </d:tag>
+    <d:tag name="heading">
+      <j:choose>
+        <j:when test="${divBasedFormLayout}">
+          <div class="th">
```
Comment (timja) [inline src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly:71]: tr is the only class that should actually on the migrated elements
```suggestion
          <div>
```

File: src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly [hunk 5]
```diff
@@ -37,10 +37,106 @@ THE SOFTWARE.
...
+          <td>
+            <d:invokeBody/>
+          </td>
+        </j:otherwise>
+      </j:choose>
+    </d:tag>
+    <d:tag name="td-colspan-3">
+      <j:choose>
+        <j:when test="${divBasedFormLayout}">
+          <div class="td">
```
Comment (timja) [inline src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly:99]: tr is the only class that should actually on the migrated elements
```suggestion
          <div>
```

File: src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly [hunk 6]
```diff
@@ -37,10 +37,106 @@ THE SOFTWARE.
...
+    </d:tag>
+    <d:tag name="td-width-100">
+      <j:choose>
+        <j:when test="${divBasedFormLayout}">
+          <div class="td">
+            <d:invokeBody/>
+          </div>
+        </j:when>
+        <j:otherwise>
+          <td width="100%">
```
Comment (timja) [inline src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly:118]: tr is the only class that should actually on the migrated elements
```suggestion
          <td>
```

File: src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly [hunk 7]
```diff
@@ -37,10 +37,106 @@ THE SOFTWARE.
...
+            <d:invokeBody/>
+          </td>
+        </j:otherwise>
+      </j:choose>
+    </d:tag>
+    <d:tag name="td-nowrap">
+      <j:choose>
+        <j:when test="${divBasedFormLayout}">
+          <div class="td">
```
Comment (timja) [inline src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly:127]: tr is the only class that should actually on the migrated elements
```suggestion
          <div>
```

File: src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly [hunk 8]
```diff
@@ -37,10 +37,106 @@ THE SOFTWARE.
...
+          </div>
+        </j:when>
+        <j:otherwise>
+          <td width="100%">
+            <d:invokeBody/>
+          </td>
+        </j:otherwise>
+      </j:choose>
+    </d:tag>
+    <d:tag name="td-nowrap">
```
Comment (timja) [inline src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly:124]: you could instead add a style one to the other td tag and just not pass it in when you don't need

For it to work in your IDE it's this I think:
`<st:attribute name="style"/>

and then `style="${attrs.style}"`

**Review by timja — APPROVED** [review]

File: README.adoc (hunk 1 above)
Comment (MarkEWaite) [reply README.adoc:197]: I worried that changing the default behavior to "disabled" might require a larger version increment and more explanation in release notes or elsewhere.  I've hesitated to change defaults of many things for fear of breaking existing use cases that were unknown to me but important to the user.

Comment (timja) [reply README.adoc:197]: if you don't change the default though will people actually discover it?

it should be a good out of the box experience for all users imo, I would just add to release notes users can opt out via configuration, and can always revert the default if people flood out not wanting this?

File: src/test/java/hudson/plugins/git/GitSCMTest.java [hunk 9]
```diff
@@ -157,6 +157,46 @@ private StandardCredentials getInvalidCredential() {
...
+    public void testHideGitTagAction() throws Exception {
+        FreeStyleProject project = setupSimpleProject("master");
+        List<UserRemoteConfig> remoteConfigs = GitSCM.createRepoList("https://github.com/jenkinsci/git-plugin", "github");
+        project.setScm(new GitSCM(remoteConfigs,
+                Collections.singletonList(new BranchSpec("master")), false, null, null, null, null));
+
+        GitSCM scm = (GitSCM) project.getScm();
+        final DescriptorImpl descriptor = (DescriptorImpl) scm.getDescriptor();
+        assertFalse("Wrong initial value for hide tag action", scm.isHideGitTagAction());
+        descriptor.setHideGitTagAction(true);
```
Comment (rishabhBudhouliya) [inline src/test/java/hudson/plugins/git/GitSCMTest.java:170]: @MarkEWaite Why are we testing the assertion on the set boolean flag two times?

Comment (MarkEWaite) [reply src/test/java/hudson/plugins/git/GitSCMTest.java:170]: Default value is tested by the first assertion, result of an explicit set(true) is tested by the second assertion.

File: README.adoc [hunk 10]
```diff
@@ -193,6 +193,19 @@ Preserve second fetch during initial checkout::
...
+Do not add git tag action to jobs::
+
+  If checked, the git tag action will not be added to any builds that happen after the box is checked.
+  The checkbox only affects builds that happen after the box is checked.
++
+The git tag action allows a user to apply a tag to the git repository in the workspace based on the git commit used in the build applying the tag.
```
Comment (MarkEWaite) [inline README.adoc:202]: This phrasing needs improvement, desperately.  The sentence is long and convoluted mess.
```suggestion
The git tag action allows a user to apply a tag to the git repository in the workspace based on the git commit used in the build applying the tag.
```

Comment (rishabhBudhouliya) [reply README.adoc:202]:
```suggestion
The git tag action applies a tag to the git repository based on the commit used in the build applying the tag. 
```
Didn't really solve the problem, I assume `git` and `repository` are already known and hence removed them.

**Review by rishabhBudhouliya — COMMENTED** [review]
@MarkEWaite The PR looks good to me, I just have one doubt. In the [tables-to-div migration ](https://www.jenkins.io/doc/developer/views/table-to-div-migration/) page I see a suggestion of using `f:entry`tag instead of `tr` or `td` elements, referring to this [work](https://github.com/jenkinsci/config-rotator-plugin/pull/2/commits/af362269dac9edf787b33e2c526cbc65bd92ac3d). I would assume it doesn't make a difference in the div layout but not sure if it is absolutely necessary to remove it from the previous layout?

File: README.adoc (hunk 1 above)
Comment (MarkEWaite) [reply README.adoc:197]: Experiments yesterday confirmed that git tag action only works consistently if the workspace is on the Jenkins controller.  Workspaces on a Jenkins agent failed to apply the requested tag in all the cases I tested.

I'll switch the default to be disabled and allow administrators to enable it.  I'll use that as the excuse to call the release of this capability "git plugin 4.5.0".  I'll include some additional comments in the documentation about the old behavior and will rework the documentation and the image to match the new default of disabling that action.

**Comment by MarkEWaite** [comment]
> @MarkEWaite The PR looks good to me, I just have one doubt. In the [tables-to-div migration ](https://www.jenkins.io/doc/developer/views/table-to-div-migration/) page I see a suggestion of using `f:entry`tag instead of `tr` or `td` elements, referring to this [work](https://github.com/jenkinsci/config-rotator-plugin/pull/2/commits/af362269dac9edf787b33e2c526cbc65bd92ac3d). I would assume it doesn't make a difference in the div layout but not sure if it is absolutely necessary to remove it from the previous layout?

I wanted to retain the old tags when running on configurations before div based layouts are available.  It was simpler to check that the behaviors are retained.

File: src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly (hunk 8 above)
Comment (MarkEWaite) [reply src/main/resources/hudson/plugins/git/GitTagAction/tagForm.jelly:124]: Considering the low usage of this action, I intentionally left it simple without additional styling.  Let me know if you find an error.

**Review by timja — APPROVED** [review]

**Review by rishabhBudhouliya — APPROVED** [review]

**Pull request merged**